README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
[![Build Status](https://travis-ci.org/shuowencs/SortedEffects.svg?branch=master)](https://travis-ci.org/shuowencs/SortedEffects)

SortedEffects
=============

The R package **SortedEffects** implements econometric methods in Chernozhukov, Fernandez-Val and Luo (2018) and reports *sorted partial effects*, a collection of estimated partial effects sorted in increasing order and indexed by percentiles. In contrast to the classical average partial effect, sorted effect curves completely represent range of heterogeneous partial effects, which are common in nonlinear predictive models. The package further allows users to conduct classification analysis with sorted effects and visualize the results. You can install the current version of with:

``` r
install.packages("devtools")
devtools::install_github("shuowencs/SortedEffects")
```

This repository hosts the (commented) source code and data. For more theoretical details, please refer to the [original paper](https://www.econometricsociety.org/publications/econometrica/2018/11/01/sorted-effects-method-discovering-heterogeneous-effects-beyond) .

This document is structured to answer three curiosity-sparked questions:

-   What do you mean by (heterogeneous) partial effects and classification analysis?
-   What function commands does **SortedEffects** offer?
-   How do I use **SortedEffects** for my research?

I will use racial-based discrimination in mortgage lending as an application to illustrate the basic functionality of **SortedEffects**. For more detailed explanations of function commands, refer to the companion vignette.

Econometrics Background
-----------------------

Many economic inquiries reduce to the following question: holding all else (control variables) fixed, how does change in a key variable affect the outcome? Depending on the assumptions we impose, such effects have different names like treatment or marginal effects. To ease conversation we call them **partial effects**. To answer the question, economists often work with nonlinear models. In particular, consider the following two kinds of nonlinearity: nonlinearity in key variables of interest or in parameters. The first kind includes OLS and quantile regressions in which the variable of interest is interacted with control variables. The second kind includes generalized linear models like Logit and Probit. It turns out that the partial effects derived under nonlinear settings vary among obervational units. We call these different effects as **heterogenous partial effects**.

Before jumping to a concrete example, let's fix some notations. Let *Y* denote outcome variable of interest and *X* = (*T*, *W*) denote regressors. In particular, *T* is the variable in interest while *W* means control variables. We link between *Y* and *X* with a predictive function *g*(*X*). With more assumptions *g* can represent causal relationship, but it's not necessary for our purposes.

The economic question is: holding *W* fixed, how do changes in *T* affect *Y*? The mathematical expression for partial effects is
*Δ*(*x*)=*g*(*t*<sub>1</sub>, *w*)−*g*(*t*<sub>0</sub>, *w*)

for discrete *T* and
*Δ*(*x*)=∂<sub>*t*</sub>*g*(*t*, *w*)
 for *T* being continuous. It turns out that if we work with nonlinear models, *Δ*(*x*) depends on values of *w*.

Here is a concrete probit example: *Y* = *Φ*(*T**β*<sub>1</sub> + *W**β*′) + *ϵ*, where *Φ* denotes CDF of standard normal distribution. If *T* is continuous, then the partial effect is
*Δ*(*x*)=*ϕ*(*T**β*<sub>1</sub> + *W**β*′)*β*<sub>1</sub>.
 It is obvious that *Δ*(*x*) depends on values of *W*. Since sample observations will have different values of *W*, *Δ*(*x*) features heterogeneity. Because *Δ*(*x*) is unobservable in data, we use econometric techniques to get an estimate. However, no estimation is accurate. Therefore we need to quantify the uncertainty of estimation, and in econometrics this is called **inference**.

It turns out hard to do inference on the estimate of partial effects, so in applied economic work researchers often report the average partial effect as an alternative (averaged over the empirical distribution of *x*). But by definition such reports neglect heterogeneity.

Chernozhukov, Fernandez-Val and Luo (forthcoming, *Econometrica*) provide theoretical underpinnings to conduct estimation and inference on the partial effect values at any quantile in interest. For example, if we want to know the partial effects from 2% to 98%, then the method will report estimate of all these effects and quantification of estimation uncertainty (confidence bands). Because the estimated partial effects are sorted in an increasing order, we call them **sorted partial effect (SPE)**.

With SPE, we can classify observational units into groups based on how greatly they are affected. For example, researchers might be curious whether mostly affected observational units have different characteristics compared to peers who are least affected.

Example 1: Racial-Based Discrimination in Mortgage Lending
----------------------------------------------------------

We use data on mortgage applications in Boston from 1990. The Federal Reserve Bank of Boston collected these data in relation to the Home Mortgage Disclosure Act (HMDA), which was passed to monitor minority access to the mortgage market. Providing better access to credit markets can arguably help the disadvantaged groups escape poverty traps. To access the data, type in the following command

``` r
data("mortgage")
```

The sample includes 2380 observations corresponding to 2041 white applicants and 339 black applicants.

We estimate a binary response model where the outcome variable *Y* is an indicator for mortgage denial, the key covariate *T* is an indicator for the applicant being black, and the controls *W* contain financial and other characteristics of the applicant that banks take into account in the mortgage decisions. The regression formula is specified as

``` r
fm <- deny ~ black + p_irat + hse_inc + ccred + mcred + pubrec + ltv_med + ltv_high + denpmi + selfemp + 
      single + hischl
```

We use `SPE` command to estimate SPE and APE.

``` r
ex <- SPE(fm = fm, data = mortgage, var.T = "black", method = "probit", us = c(2:98)/100, B = 200)
```

In this line of code, `var.T = "black"` tells the package that *T* is `black`, while `us` specifies the SPE quantile index to range from 2% to 98%. `B` refers to the number of bootstrap for inference. For accuracy we recommend setting `B` to be 500. The result `ex` is a list containing SPE and APE with corresponding confidence bands, and for visualization we can use the `plot.SPE` command as follows:

``` r
plot.SPE(SE=ex$spe, AE=ex$ape, us=c(2:98)/100, alpha=0.1, main="APE and SPE of Being Black on the prob of 
         Mortgage Denial", sub="Probit Model", ylab="Change in Probability")
```

We see that the SPE ranges from 2% to 14% while APE is 6%. It's clear that reporting APE masks rich heterogeneity of the partial effects.

Now that we know being black affects mortgage denial probability differently among applicants, we can ask a follow-up question: what are the average characteristics of most and least affected applicants? Such questions are referred to as **classification analysis**: first classify observational units into most/least affected groups and then study objects of interest like mean and distributions.

To conduct estimation and inference on classification analysis, we use the `u.CA` command. We first choose the characteristics in interest: Deny, Black, Debt-to-income, Expenses-to-income, Bad consumer credit, Bad mortgage credit, Credit problems, Denied mortgage insurance, Medium loan-to-house, High loan-to-house, Self employed, Single and High school grad. The corresponding specification is

``` r
t <- c(rep(1, 4), 0, rep(1, 7), 0, 0, 1, 1)
```

Then we issue to `u.CA` command:

``` r
CA <- u.CA(fm=fm, data=mortgage, var.T="black", method="probit",
           cl=matrix(c(1,0,0,1), nrow=2), t=t, B=200)
```

The specification `cl=matrix(c(1,0,0,1), nrow=2)` means we tabulate the mean of the characteristics for two groups. The result `CA` is a list containing the estimate, standard errors and joint p-values. We can tabulate the result as follows:

``` r
est <- matrix(CA$est, ncol=2)
se <- matrix(CA$bse, ncol=2)
Table <- matrix(0, ncol=4, nrow=13)
Table[, 1] <- est[, 1] # Least Affected Bias-corrected estimate
Table[, 2] <- se[, 1] # Corresponding SE
Table[, 3] <- est[, 2] # Most affected
Table[, 4] <- se[, 2] # Corresponding SE
rownames(Table) <- colnames(CA$est)[1:13] # assign names to each row
colnames(Table) <- rep(c("Estimate", "SE"), 2)
xtable(Table)
```

We can see that the characteristics are different for these two groups.


NEWS.md
NEWS
================
Shuowen Chen
11/4/2019

# SortedEffects 1.1.0.
This note document the changes from SortedEffects 1.0.0. The following changes are suggested by Norman Martloff, the R Journal editor, Thomas Leeper and an anonymous referee from the R Journal. 

## Major Changes
1. Compared to the previous version 1.0.0, the package now has a S3 structure. The plotting functions (`plot.spe`, `plot.ca` and `plot.subpop`) now are methods of generic `plot()`. 

2. In addition to the plotting methods, we provide three new summary methods (`summary.spe`, `summary.ca` and `summary.subpop`), which are methods of generic `summary()`. We believe these three new methods provide richer outputs to the users. 

## Minor Changes
1. We decaptalize the three functions (`spe`, `ca`, `subpop`). 
2. We rename some arguments in the three functions to be snake case rather than dot case. These are `var` (instead of `var.T`), `boot_type` (instead of `boot.type`), `range_cb` (instead of `range.cb`), `var_type` (instead of `var.type`). We rename the argument `B` in the three functions to be `b`, and change the default to 500. 
3. The arguments `method`, `boot_type`, `var_type` are have a finite set of options for users to choose from. So we now explicitly provide these options in the argument, and the code will use `match.arg()` to match user input for execution. 
4. In each of the three functions (`spe`, `ca`, `subpop`) we add an option called `parallel` that allows users to either turn on or off the parallel computing. If users turn on the parallel computing, they can further specify how many CPUs to use via the `cores` option. 
5. The package now has a progress bar showing the progress of bootstrap estimation. It also shows how many CPUs the users are using. 
6. The option `cl` in function `ca` wasn't very intuitive. Now the users can choose from two alternatives (`diff` or `hoth`), which are the most common for `cl`. 
7. In `plot.subpop`, we add an option called `overlap` that allow users to either keep or drop the overlapped observations in the plot. 
8. In the previous version, the option `t` and `cat` in function `ca` required users to manually input all the variables. In particular, if some variables of interest are factor, the users need to manually create a lot of indicator variables and type them as an input. In the current version, users only need to input the factor variable itself, and the code will automatically generate indicator variables and run the estiamtions. We believe this makes the package easier to use. 
9. In each function (`spe`, `ca`, `subpop`), we change the default of option `taus` to `c(5:95)/100`. 
10 In function `ca`, we change the default to `c(1:99)/100`. 


R/data.R
#' Mortgage Denial
#'
#' @format Contains the data on mortgage application in Boston from 1990,
#' (Munnell et al., 1996.)
#' We obtain the data from the companion website of Stock and Watson (2011).
#' The file contains the following variables:
#' \describe{
#' \item{deny}{indicator for mortgage application denied}
#' \item{p_irat}{monthly debt to income ratio}
#' \item{black}{indicator for black applicant}
#' \item{hse_inc}{monthly housing expenses to income ratio}
#' \item{loan_val}{loan to assessed property value ratio}
#' \item{ccred}{consumer credit score with 6 categories. 1 if no "slow"
#' payments or delinquencies, 2 if one or two "slow" payments or delinquencies,
#' 3 if more than two "slow" payments or delinguencies, 4 if insufficient
#' credit history for determination, 5 if delinquent credit history with
#' payment 60 days overdue, and 6 if delinquent credit history with payments
#' 90 days overdue.}
#' \item{mcred}{mortgage credit score with 4 categories. 1 if no late mortgage
#' payments, 2 if no mortgage payment history, 3 if one or two late mortage
#' payments, and 4 if more than two late mortgages payments}
#' \item{pubrec}{indicator for any public record of credit problems: bankruptcy
#' , charge-offs, collection actions}
#' \item{denpmi}{indicator for applicant applied for mortgage insurance and was
#' denied}
#' \item{selfemp}{indicator for self-employed applicant}
#' \item{single}{indicator for single applicant}
#' \item{hischl}{indicator for high school graduated applicant}
#' \item{probunmp}{1989 Massachusetts unemployment rate in the applicant's
#' history}
#' \item{condo}{indicator for unit is a condominium}
#' \item{ltv_med}{indicator for medium loan to property value ratio [.80, .95]}
#' \item{ltv_high}{indicator for high loan to property value ratio >.95}
#'}
#'@source Munnell, Alicia, Geoffrey Tootell, Lynn Browne, and James McEneaney,
#'"Mortgage Lending in Boston: Interpreting HMDA Data", The American Economic
#'Review, 1996.
"mortgage"

#' Wage Data
#'
#' @format Consists of white, non-hispanic individuals aging from 25 to 64 and
#' working more than 35 hours per week during at least 50 weeks of the year.
#' Excludes self-employed, individuals living in group quarters; individuals in
#' the military, agricultural or private household sectors; individuals with
#' inconsistent reports on earnings and employment status; individuals with
#' allocated or missing information in any of the variables used in the
#' analysis; individuals with hourly wage rate below $3. Contains 32,523
#' workers including 18,137 men and 14,382 women. The file contains the
#' following variables:
#' \describe{
#' \item{wage}{hourly wages}
#' \item{lnw}{log of hourly wages}
#' \item{widowed}{marital status indicator: widowed}
#' \item{divorced}{marital status indicator: divorced}
#' \item{separated}{marital status indicator: separatd}
#' \item{nevermarried}{marital status indicator: nevermarried}
#' \item{married}{marital status indicator: married}
#' \item{weight}{CPS sampling weight}
#' \item{female}{gender indicator: 1 if female}
#' \item{lhs}{education attainment indicator: less than high school graduate:
#' years of educ < 12}
#' \item{hsg}{education attainment indicator:high school graduate: years of
#' educ = 12}
#' \item{sc}{education attainment indicator: some college: 13<=years of
#' educ<=15}
#' \item{cg}{education attainment indicator: college: 16<=years of educ<=17}
#' \item{ad}{education attainment indicator: advanced degree: years of
#' educ>=18}
#' \item{mw}{region indicator: midwest}
#' \item{so}{region indicator: south}
#' \item{we}{region indicator: west}
#' \item{me}{region indicator: northeast}
#' \item{exp1}{max(age-years of educ-7, 0)}
#' \item{exp2}{exp1^2/100}
#' \item{exp3}{exp1^3/100}
#' \item{exp4}{exp1^4/100}
#' \item{occ}{Occupation Categories in CPS: 456 categories}
#' \item{occ2}{Aggregated occupation: 6 categories, 5 categories without armed
#' forces}
#' \item{ind}{Industry Categories in CPS: 257 categories}
#' \item{ind2}{Aggregated Industry: 14 categories, 12 without agriculture and
#' military}
#' }
#'
#' @source U.S. March Supplement of the Current Population Survey (CPS) in 2015.
#'
"wage2015"


R/spe.R
#' Empirical Sorted Partial Effects (SPE) and Inference
#'
#' \code{spe} conducts SPE estimation and inference at user-specifed quantile
#' index. The bootstrap procedures follows algorithm 2.1 as in Chernozhukov,
#' Fernandez-Val and Luo (2018). All estimates are bias-corrected and all
#' confidence bands are monotonized. For graphical results, please use
#' \code{\link{plot.spe}}.
#'
#' @return The output is a list with 4 components: (1) \code{spe} stores spe
#' estimates, the upper and lower confidence bounds, and standard errors;
#' (2) \code{ape} stores ape estimates, the upper and lower confidence bounds,
#' and the standard error; (3) \code{us} stores percentile index as in \
#' code{spe} command; (4) \code{alpha} stores significance level as in
#' \code{spe} command.
#'
#' @param fm          Regression formula.
#' @param data        Data in use.
#' @param method      Models to be used for estimating partial effects. Four
#'                    options: \code{"logit"} (binary response),
#'                    \code{"probit"} (binary response), \code{"ols"}
#'                    (interactive linear with additive errors), \code{"QR"}
#'                    (linear model with non-additive errors). Default is
#'                    \code{"ols"}.
#' @param var_type    The type of parameter in interest. Three options:
#'                    \code{"binary"}, \code{"categorical"},
#'                    \code{"continuous"}. Default is \code{"binary"}.
#' @param var         Variable T in interset. Should be a character type.
#' @param compare     If parameter in interest is categorical, then user needs
#'                    to specify which two category to compare with. Should be
#'                    a 1 by 2 character vector. For example, if the two levels
#'                    to compare with is 1 and 3, then \code{c=("1", "3")},
#'                    which will calculate partial effect from 1 to 3. To use
#'                    this option, users first need to specify \code{var} as a
#'                    factor variable.
#' @param subgroup    Subgroup in interest. Default is \code{NULL}.
#'                    Specifcation should be a logical variable. For example,
#'                    suppose data contains indicator variable for women
#'                    (female if 1, male if 0). If users are interested in
#'                    women SPE, then users should specify \code{subgroup =
#'                    data[, "female"] == 1}.
#' @param samp_weight Sampling weight of data. Input should be a n by 1 vector,
#'                    where n denotes sample size. Default is \code{NULL}.
#' @param us          Percentile of interest for SPE. Should be a vector of
#'                    values between 0 and 1. Default is \code{c(1:9)/10}.
#' @param alpha       Size for confidence interval. Shoule be between 0 and 1.
#'                    Default is 0.1
#' @param taus        Indexes for quantile regression. Default is
#'                    \code{c(5:95)/100}.
#' @param b           Number of bootstrap draws. Default is set to be 500.
#' @param parallel    Whether the user wants to use parallel computation.
#'                    The default is \code{FALSE} and only 1 CPU will be used.
#'                    The other option is \code{TRUE}, and user can specify
#'                    the number of CPUs in the \code{ncores} option.
#' @param ncores      Number of cores for computation. Default is set to be
#'                    \code{detectCores()}, which is a function from package
#'                    \code{parallel} that detects the number of CPUs on the
#'                    current host. For large dataset, parallel computing is
#'                    highly recommended since bootstrap is time-consuming.
#' @param seed        Pseudo-number generation for reproduction. Default is 1.
#' @param bc          Whether want the estimate to be bias-corrected. Default
#'                    is \code{TRUE}. If \code{FALSE} uncorrected estimate and
#'                    corresponding confidence bands will be reported.
#' @param boot_type   Type of bootstrap. Default is \code{"nonpar"}, and the
#'                    package implements nonparametric bootstrap. The other
#'                    alternative is \code{"weighted"}, and the package
#'                    implements weighted bootstrap.
#' @examples
#' data("mortgage")
#' fm <- deny ~ black + p_irat + hse_inc + ccred + mcred + pubrec + ltv_med +
#' ltv_high + denpmi + selfemp + single + hischl
#'
#' test <- spe(fm = fm, data = mortgage, var = "black", method = "logit",
#' us = c(2:98)/100, b = 50)
#'
#' @importFrom Hmisc wtd.quantile
#' @importFrom boot boot
#' @importFrom stats quantile rexp qnorm
#' @importFrom parallel detectCores
#' @importFrom pbapply setpb startpb closepb
#' @export
spe <- function(fm, data, method = c("ols", "logit", "probit", "QR"),
                var_type = c("binary", "continuous", "categorical"), var,
                compare, subgroup = NULL, samp_weight = NULL, us = c(1:9)/10,
                alpha = 0.1, taus = c(5:95)/100, b = 500, parallel = FALSE,
                ncores = detectCores(), seed = 1, bc = TRUE,
                boot_type = c("nonpar", "weighted")) {
  # ----- Stopping Condition
  if (alpha >= 1 || alpha <= 0) stop("Please specify a correct size for
                                     hypothesis testing between 0 and 1.")
  # ----- Replace Null samp_weight Specification
  if (is.null(samp_weight)) samp_weight <- rep(1, dim(data)[1])
  # ----- Matching Arguments
  method <- match.arg(method)
  var_type <- match.arg(var_type)
  boot_type <- match.arg(boot_type)
  # ----- 1. Call to estimate PE
  output <- suppressWarnings(peestimate(fm, data, samp_weight, var_type, var, compare, method,
                       subgroup, taus))
  pe_est <- output$pe_est
  # ----- 2. Now get estimated SPE (full and subgroup samples)
  if (method != "QR") {
    spe_est <- wtd.quantile(pe_est, samp_weight, us)
  } else {
    spe_est <- wtd.quantile(pe_est, matrix(samp_weight, ncol = 1,
                                           nrow = nrow(pe_est),
                                           byrow = FALSE), us)
  }
  if (!is.null(subgroup)) {
    pesub_est <- output$pesub_est
    pesub_w <- output$samp_weight_sub
    spesub_est <- wtd.quantile(pesub_est, pesub_w, us)
  }
  # -----3. Bootstrap Samples
  #  statistic in one bootstrap for SPE
  #  Need to consider two scenarios: whether the regression is weighted. If not
  #  then use nonpar boot, if yes use exchangeably weighted boot. Turns out the
  #  statistic functions for boot are a bit different for these two scenarios,
  #  so I write two stastistics functions.

  # set a bootstrap counting variable for the purpose of showing a progress bar
  rep_count <- 1
  stat_boot_noweight <- function(data, indices) {
    data$.w <- samp_weight
    # create bootstrap sample, the "indices" is automatically provided by the
    # boot package, provides the sampling with replacement part of bootstrap
    data_bs <- data[indices, ]
    # set up a progress bar to document the bootstrap progress
    setpb(pb, rep_count)
    rep_count <<- rep_count + 1
    output_bs <- peestimate(fm, data_bs, samp_weight = data_bs$.w, var_type,
                            var, compare, method, subgroup, taus)
    est_pe_bs <- output_bs$pe_est
    # scalar, length = 1
    est_ape_bs <- mean(est_pe_bs)
    # length = length(us)
    est_spe_bs <- wtd.quantile(est_pe_bs, data_bs$.w, us)
    if (!is.null(subgroup)) {
      est_pesub_bs <- output_bs$pesub_est
      pesub_w_bs <- output_bs$samp_weight_sub
      # length = 1
      est_apesub_bs <- mean(est_pesub_bs)
      # length(us)
      est_spesub_bs <- wtd.quantile(est_pesub_bs, pesub_w_bs, us)
      # Each has length: 1+length(us)
      return(c(est_apesub_bs, est_spesub_bs))
    } else {
      return(c(est_ape_bs, est_spe_bs))
    }
    data_bs$.w <- NULL
  }
  # The resampling is wrt sample weight, so the data_rg function does that
  data_rg <- function(data, mle) {
    n <- dim(data)[1]
    # Exponential weights
    multipliers  <- rexp(n)
    # Sampling weight of data.bs
    weight <- (multipliers/sum(multipliers)) * samp_weight
    data$.w <- weight
    return(data)
  }
  # Implements nonparametric bootstrap for quantile regression
  data_non <- function(data, mle) {
    n <- dim(data)[1]
    multipliers <- as.vector(table(factor(sample(n,n,replace = T),
                                          levels = c(1:n))))
    # Sampling weight of data.bs
    weight <- (multipliers/sum(multipliers)) * samp_weight
    data$.w <- weight
    return(data)
  }
  # The following is boot stat function for exchageably weighted boot
  stat_boot_weight <- function(data) {
    setpb(pb, rep_count)
    rep_count <<- rep_count + 1
    output_bs <- peestimate(fm, data, samp_weight = data$.w, var_type, var,
                            compare, method, subgroup, taus)
    est_pe_bs <- output_bs$pe_est
    est_ape_bs <- mean(est_pe_bs) # scalar, length = 1
    if (method != "QR") {
      est_spe_bs <- wtd.quantile(est_pe_bs, samp_weight, us)
    } else {
      est_spe_bs <- wtd.quantile(est_pe_bs, matrix(samp_weight, ncol = 1,
                                                   nrow = nrow(pe_est),
                                                   byrow = FALSE), us)
    } # length = length(us)
    if (!is.null(subgroup)) {
      est_pesub_bs <- output_bs$pesub_est
      pesub_w_bs <- output_bs$samp_weight_sub
      est_apesub_bs <- mean(est_pesub_bs) # length = 1
      est_spesub_bs <- wtd.quantile(est_pesub_bs, pesub_w_bs, us) # length(us)
      return(c(est_apesub_bs, est_spesub_bs)) # Each has length: 1+length(us)
    } else {
      return(c(est_ape_bs, est_spe_bs))
    }
  }
  # Use boot command
  set.seed(seed)
  if (parallel == FALSE) ncores <- 1
  if (boot_type == "nonpar") {
    if (method != "QR") {
      # print a message showing how many cores are used
      cat(paste("Using", ncores, "CPUs now.\n"))
      # set up a progress bar
      pb <- startpb(min = 0, max = b)
      result_boot <- boot(data = data, statistic = stat_boot_noweight,
                          parallel = "multicore", ncpus = ncores, R = b)
      closepb(pb)
    } else {
      data$.w <- samp_weight
      cat(paste("Using", ncores, "CPUs now.\n"))
      pb <- startpb(min = 0, max = b)
      result_boot <- boot(data = data, statistic = stat_boot_weight,
                          sim = "parametric", ran.gen = data_non, mle = 0,
                          parallel = "multicore", ncpus = ncores, R = b)
      data$.w <- NULL
      closepb(pb)
    }
  } else if (boot_type == "weighted") {
    data$.w <- samp_weight
    cat(paste("Using", ncores, "CPUs now.\n"))
    pb <- startpb(min = 0, max = b)
    result_boot <- boot(data = data, statistic = stat_boot_weight,
                        sim = "parametric", ran.gen = data_rg, mle = 0,
                        parallel = "multicore", ncpus = ncores, R = b)
    data$.w <- NULL
    closepb(pb)
  }
  # ------ 4. Inference
  ##############################
  ### Full Sample
  ##############################
  # SPE inference
  draws_spe_bs <- result_boot$t[, 2:(length(us) + 1)]
  # bundle function implements algorithm 2.1, remark 2.2 and 2.3
  inf_spe <- bundle(draws_spe_bs, spe_est, alpha)
  # APE inference
  est_ape <- mean(pe_est)
  draws_ape_bs <- result_boot$t[, 1]
  inf_ape <- ape(draws_ape_bs, est_ape, alpha)
  ##############################
  ### Subgroup Sample (if subgroup is not NULL)
  ##############################
  if (!is.null(subgroup)) {
    # SPE
    draws_spesub_bs <- result_boot$t[, 2:(length(us) + 1)]
    inf_spesub <- bundle(draws_spesub_bs, spesub_est, alpha)
    # APE
    est_apesub <- mean(pesub_est)
    draws_apesub_bs <- result_boot$t[, 1]
    inf_apesub <- ape(draws_apesub_bs, est_apesub, alpha)
  }
  # ----- 5. Return Results
  # Depends on whether subgroup is NULL and whether bias correction is wanted
  # The resulting outputs are APE(sub) & SPE(sub) with correpsonding confidence
  # intervals. All are stored in an "inf" bundle list
  if (!is.null(subgroup)) {
    if (bc == TRUE) {
      output <- list(spe = inf_spesub[c(1:3, 7)], ape = inf_apesub[c(1:3, 7)],
                     us = us, alpha = alpha)
    } else {
      output <- list(spe = inf_spesub[4:7], ape = inf_apesub[4:7], us = us,
                     alpha = alpha)
    }
  } else {
    if (bc == TRUE) {
      output <- list(spe = inf_spe[c(1:3, 7)], ape = inf_ape[c(1:3, 7)],
                     us = us, alpha = alpha)
    } else {
      output <- list(spe = inf_spe[4:7], ape = inf_ape[4:7], us = us,
                     alpha = alpha)
    }
  }
  # claim output as an object of class "spe" so as to use S3
  output <- structure(output, class = "spe")
  return(output)
}

# ------- Auxiliary Functions ----------------
# Implementing algorithm 2.1 and get (bias corrected) estimate and confidence
# bands for sorted effects
bundle <- function(bs, est, alpha) {
  z_bs <- bs - matrix(est, nrow = nrow(bs), ncol = ncol(bs), byrow = TRUE)
  sigma <- (apply(z_bs, 2, quantile, .75, na.rm = TRUE) -
              apply(z_bs, 2, quantile, .25, na.rm = TRUE)) / (qnorm(0.75) -
                                                              qnorm(.25))
  t_hat <- apply(abs(z_bs / matrix(sigma, nrow = nrow(bs), ncol = ncol(bs),
                                   byrow = TRUE)), 1, max)
  crt <- quantile(t_hat, 1 - alpha)
  # bias-correction
  est_bc <- sort(2*est - apply(bs, 2, mean))
  ubound_est_bc <- sort(est_bc + crt * sigma)
  lbound_est_bc <- sort(est_bc - crt * sigma)
  # uncorrected
  ubound_est <- sort(est + crt * sigma)
  lbound_est <- sort(est - crt * sigma)
  out <- list(est_bc = est_bc, ubound_est_bc = ubound_est_bc,
              lbound_est_bc = lbound_est_bc, est = est, ubound_est = ubound_est,
              lbound_est = lbound_est, sigma_spe = sigma)
  return(out)
}

# Get (biased corrected) average partial effects and confidence intervals
ape <- function(bs, est, alpha) {
  sigma <- (quantile(bs, .75, na.rm = TRUE) -
              quantile(bs, .25, na.rm = TRUE)) / (qnorm(0.75) - qnorm(.25))
  mzs <- abs(bs - est) / sigma
  crt2 <- quantile(mzs, 1 - alpha)
  # bias-correction
  est_bc <- 2 * est - mean(bs)
  ubound_ape_bc <- est_bc + crt2 * sigma
  lbound_ape_bc <- est_bc - crt2 * sigma
  # uncorrected
  ubound_ape <- est + crt2 * sigma
  lbound_ape <- est - crt2 * sigma
  out <- list(est_bc = est_bc, ubound_est_bc = ubound_ape_bc,
              lbound_est_bc = lbound_ape_bc, est = est, ubound_ape = ubound_ape,
              lbound_ape = lbound_ape, sigma_ape = sigma)
  return(out)
}

# ------- Plotting Function ----------------
# Since plot() is a generic, we write a new plotting method for the SPE class.
# --Plotting (Output: average effect, sorted effect, correspondent confidence bands) --
#' Plot output of \code{\link{spe}} command. The x-axis limits are set to the
#' specified range of percentile index.
#' @param x           Output of \code{\link{spe}} command.
#' @param ylim        y-axis limits. Default is NULL.
#' @param main        Main title of the plot. Defualt is NULL.
#' @param sub         Sub title of the plot. Default is NULL.
#' @param xlab        x-axis label. Default is "Percentile Index".
#' @param ylab        y-axis label. Default is "Sorted Effects".
#' @param ...         graphics parameters to be passed to the plotting
#'                    routines.
#' @examples
#' data("mortgage")
#' fm <- deny ~ black + p_irat + hse_inc + ccred + mcred + pubrec + ltv_med +
#' ltv_high + denpmi + selfemp + single + hischl
#' test <- spe(fm = fm, data = mortgage, var = "black", method = "logit",
#' us = c(2:98)/100, b = 50)
#'
#' plot(x = test, main="APE and SPE of Being Black on the prob of
#' Mortgage Denial", sub="Logit Model", ylab="Change in Probability")
#'
#' @importFrom graphics plot polygon lines abline points legend
#' @export
plot.spe <- function(x, ylim = NULL, main = NULL, sub = NULL,
                     xlab = "Percentile Index", ylab = "Sorted Effects",
                     ...) {
  # SPE and confidence bands
  SE <- x$spe
  AE <- x$ape
  us <- x$us
  alpha <- x$alpha
  xlim <- range(us)
  plot(us, SE[[1]], type = "l", xlim, ylim, log = "", main, sub, xlab, ylab,
       col = 4, lwd = 2)
  polygon(c(us, rev(us)),c(SE[[2]], rev(SE[[3]])), density = 60, border = F,
          col = 'light blue', lty = 1, lwd = 1)
  lines(us, SE[[1]], lwd = 2, col = 4 )
  # APE and CI
  abline(h = AE[[1]], col = 1)
  abline(h = AE[[2]], col = 1, lty = 2)
  abline(h = AE[[3]], col = 1, lty = 2)
  points(c(min(us),.2,.4,.6,.8,max(us)), rep(AE[[1]], 6), col = 1, pch = 15)
  legend(x = "topleft", col = c(4, 1, "light blue", 1), lwd = c(1, 1, 5, 1),
         lty = c(1, 1, 1, 2), pch = c(NA, 15, NA, NA), pt.cex = c(2, 1),
         bty = 'n', legend = c("SPE","APE", paste0((1 - alpha)*100,"% CB(SPE)"),
                    paste0((1 - alpha)*100,"% CB(APE)")))
}

# ------- Summary Function ----------------
#' Tabulate the output of \code{\link{spe}} function.
#'
#' The option \code{result} allows user to tabulate either sorted estimates or
#' average estimates. For sorted estimates, the table shows user-specified
#' quantile indices, sorted estimates, standard errors, point-wise confidence
#' intervals, and uniform confidence intervals. For average estimates, the
#' table shows average estiamtes, standard errors, and confidence intervals.
#'
#' @param object   The output of \code{\link{spe}} function.
#' @param result   Whether the user wants to see the sorted or the average
#'                 estimates. Default is \code{sorted}, which shows the
#'                 sorted estimates.
#' @param ...      additional arguments affecting the summary produced.
#' @examples
#' data("mortgage")
#' fm <- deny ~ black + p_irat + hse_inc + ccred + mcred + pubrec + ltv_med +
#' ltv_high + denpmi + selfemp + single + hischl
#' test <- spe(fm = fm, data = mortgage, var = "black", method = "logit",
#' us = c(2:98)/100, b = 50)
#' summary(test)
#' @export
summary.spe <- function(object, result = c("sorted", "average"), ...) {
  spe <- object$spe
  ape <- object$ape
  us <- object$us
  alpha <- object$alpha
  result <- match.arg(result)
  if (result == "sorted") {
    table <- matrix(0, nrow = length(us), ncol = 6)
    table[, 1] <- unlist(spe[1])
    table[, 2] <- unlist(spe[4])
    table[, 3] <- table[, 1] - qnorm(1 - alpha/2)*table[, 2]
    table[, 4] <- table[, 1] + qnorm(1 - alpha/2)*table[, 2]
    table[, 5] <- unlist(spe[3])
    table[, 6] <- unlist(spe[2])
    rownames(table) <- us
    colnames(table) <- c("Est", "SE", paste0((1 - alpha)*100,"% PLB"),
                         paste0((1 - alpha)*100,"% PUB"),
                         paste0((1 - alpha)*100,"% ULB"),
                         paste0((1 - alpha)*100,"% UUB"))
  } else {
    table <- matrix(0, nrow = 1, ncol = 4)
    table[, 1] <- unlist(ape[1])
    table[, 2] <- unlist(ape[4])
    table[, 3] <- unlist(ape[3])
    table[, 4] <- unlist(ape[2])
    rownames(table) <- "APE"
    colnames(table) <- c("Est", "SE", paste0((1 - alpha)*100,"% LB"),
                         paste0((1 - alpha)*100,"% UB"))
  }
  return(table)
}
